era/front.js

```diff
@@ -131,3 +150,508 @@ function setParamAndReload() {
+        selectEnemy(config);
+      }
+    );
+  plotXRange = [0, (fight.end - fight.start) / 1000];
+  globalThis.Plotly.newPlot(el_plot, plotData, {
+    title: `Threat - ${enemy.name}`,
+    titlefont: { color: "#fff" },
+    xaxis: {
+      title: "Time (s)",
+      titlefont: { color: "#fff" },
+      tickcolor: "#666",
+      tickfont: { color: "#fff" },
+      rangemode: "tozero",
+      gridcolor: "#666",
+      linecolor: "#999",
+      range: plotXRange.slice(),
+    },
+    yaxis: {
+      title: "Threat",
+      titlefont: { color: "#fff" },
+      tickcolor: "#666",
+      tickfont: { color: "#fff" },
+      rangemode: "tozero",
+      gridcolor: "#666",
+      linecolor: "#999",
+    },
+    width: window.innerWidth - SCROLLBAR_WIDTH,
+    height: (window.innerWidth - SCROLLBAR_WIDTH) / (1920 / 1080),
+    hovermode: "closest",
+    plot_bgcolor: "#222",
+    paper_bgcolor: "#222",
+    legend: { font: { color: "#fff" } },
+  });
+  el_plot.on("plotly_click", (/** @type {{ points: string | any[]; }} */ e) => {
+    if (e.points.length === 0) return;
+    selectTarget(
+      config,
+      reportId +
+        ";" +
+        fight.id +
+        ";" +
+        enemy.key +
+        ";" +
+        e.points[0].data.unitKey
+    );
+  });
+  el_plot.on("plotly_relayout", (/** @type {{ [x: string]: any; }} */ e) => {
+    let range = [e["xaxis.range[0]"], e["xaxis.range[1]"]];
+    if (range[0] === undefined || range[1] === undefined) return;
+    if (range[0] === plotXRange[0] && range[1] === plotXRange[1]) return;
+    plotXRange = range;
+    selectTarget(config);
+  });
+}
+
+/**
+ * @param {HTMLDivElement} el_div
+ * @param {ThreatTrace} trace
+ */
+function tabulateThreat(el_div, trace) {
+  let el_table = document.createElement("table");
+
+  function addRow(texts) {
+    let el_tr = document.createElement("tr");
+    for (let i = 0; i < texts.length; ++i) {
+      let el_td = document.createElement("td");
+      if (typeof texts[i] === "number") {
+        el_td.textContent = texts[i].toFixed(2);
+        el_td.align = "right";
+      } else {
+        el_td.textContent = texts[i];
+      }
+      el_tr.appendChild(el_td);
+    }
+    el_table.appendChild(el_tr);
+  }
+
+  let rangeWidth = plotXRange[1] - plotXRange[0];
+  el_table.innerHTML = `<tr><th>Ability</th><th title="Over the currently zoomed x range.">Threat (*)</th><th>Per ${rangeWidth.toFixed(
+    2
+  )} seconds</th></tr>`;
+  let a = trace.threatBySkill(plotXRange);
+  let keys = Object.keys(a);
+  keys.sort((x, y) => a[y] - a[x]);
+  let totalThreat = 0;
+  for (let i = 0; i < keys.length; ++i) {
+    let k = keys[i];
+    totalThreat += a[k];
+    addRow([k, a[k], a[k] / rangeWidth]);
+  }
+  addRow(["Total", totalThreat, totalThreat / rangeWidth]);
+  el_div.appendChild(el_table);
+}
+
+/**
+ * @param {import("./base.js").GameVersionConfig} config
+ * @param {HTMLDivElement} el_div
+ * @param {ThreatTrace} trace
+ */
+function tabulate(config, el_div, trace) {
```

Comment:
Same re `tabulate(el_div)`, moved out to `tabulate(config, el_div, trace)` to  keep UI code separate.